harvester/create.go

```diff
@@ -0,0 +1,202 @@
+package harvester
+
+import (
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/harvester/go-harvester/pkg/builder"
+	goharverrors "github.com/harvester/go-harvester/pkg/errors"
+	"github.com/rancher/machine/libmachine/drivers"
+	"github.com/rancher/machine/libmachine/log"
+	"github.com/rancher/machine/libmachine/mcnutils"
+	"github.com/rancher/machine/libmachine/ssh"
+	"github.com/rancher/machine/libmachine/state"
+)
+
+const (
+	keypairNamespace = "harvester-system"
+)
+
+func (d *Driver) PreCreateCheck() error {
+	c, err := d.getClient()
+	if err != nil {
+		return err
+	}
+
+	// vm is not exist
+	_, err = c.VirtualMachines.Get(d.Namespace, d.MachineName)
+	if err == nil {
+		return fmt.Errorf("machine %s is already exist", d.MachineName)
+	}
+
+	// image exist
+	image, err := c.Images.Get(d.Namespace, d.ImageName)
+	if err != nil {
+		if goharverrors.IsNotFound(err) {
+			return fmt.Errorf("image %s is not exist in namespace %s", d.ImageName, d.Namespace)
+		}
+		return err
+	}
+
+	// image succeed
+	if image.Status.Progress != 100 {
+		return fmt.Errorf("image %s's progress %d != 100", image.Name, image.Status.Progress)
+	}
+	d.ImageDownloadURL = image.Status.DownloadURL
+
+	if d.KeyPairName != "" {
+		keypair, err := c.Keypairs.Get(keypairNamespace, d.KeyPairName)
+		if err != nil {
+			if goharverrors.IsNotFound(err) {
+				return fmt.Errorf("keypair %s is not exist in namespace %s", d.KeyPairName, keypairNamespace)
+			}
+			return err
+		}
+
+		// keypair validated
+		keypairValidated := false
+		for _, condition := range keypair.Status.Conditions {
+			if condition.Type == "validated" && condition.Status == "True" {
+				keypairValidated = true
+			}
+		}
+		if !keypairValidated {
+			return fmt.Errorf("keypair %s is not validated", keypair.Name)
+		}
+
+		d.SSHPublicKey = keypair.Spec.PublicKey
+	}
+
+	// network exist
+	if d.NetworkType != networkTypePod {
+		_, err = c.Networks.Get(d.Namespace, d.NetworkName)
+		if err != nil {
+			if goharverrors.IsNotFound(err) {
+				return fmt.Errorf("network %s is not exist in namespace %s", d.KeyPairName, d.Namespace)
+			}
+			return err
+		}
+	}
+
+	return err
+}
+
+func (d *Driver) Create() error {
+	c, err := d.getClient()
+	if err != nil {
+		return err
+	}
+
+	if err = d.createKeyPair(); err != nil {
+		return err
+	}
+	userData, networkData := d.createCloudInit()
+	// create vm
+	vmBuilder := builder.NewVMBuilder("docker-machine-driver-harvester").
+		Namespace(d.Namespace).Name(d.MachineName).
+		CPU(d.CPU).Memory(d.MemorySize).
+		Image(d.DiskSize, d.DiskBus, d.ImageDownloadURL).
+		CloudInit(userData, networkData)
+
+	if d.KeyPairName != "" {
+		vmBuilder = vmBuilder.SSHKey(d.KeyPairName)
+	}
+
+	if d.NetworkType != networkTypePod {
+		vmBuilder = vmBuilder.Bridge(d.NetworkName, d.NetworkModel)
+	} else {
+		vmBuilder = vmBuilder.ManagementNetwork(true)
```

Comment:
Can you clarify the reason for using a bridge network as the default? Is it possible that we inject some iptables rules so that defaulting to masquerade can work?

Reply:
Druring rke install a cluster, rke performs port connectivity checks between cluster nodes
1. etcd <-> etcd
2. control plane <-> etcd
3. control plane <-> workers
If node use masquerade mode, tcp://pod ip:port is not accessible from inside the virtual machine.
Although we can add nat rules such as `iptables -t nat -A OUTPUT -d 10.42.2.21/32 -p tcp -m tcp -j DNAT --to-destination 127.0.0.1` inside the virtual machine to make the port accessible, but since the added nat rule contains the pod's ip and the pod ip will change after live migration, the nat rule added inside the VM needs to be updated, which is difficult to implement in a good way. 
If the pod ip remains the same during live migration, then it will cause an ip conflict.

Reply:
quick question, how could we get internet access inside deployed Kubernetes cluster using masquerade (on VM network we don't have internet access and we would like to use the mgmt interface from each node to have internet access)?

Thank you!

Reply:
My plan would be to have two VM interfaces (the first one on Bridge mode on VLAN10 and the second one masquerade on mgmt interface). Would it work this way?